buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -314,110 +308,282 @@ void reallocate(int size, int maxCapacity, AdaptiveByteBuf into) {
+                assert allocated : "Allocation of threadLocalMagazine must always succeed";
+                return buf;
+            }
+
+            // Path for concurrent allocation.
+            long threadId = currentThread.getId();
+            Magazine[] mags;
+            int expansions = 0;
+            do {
                 mags = magazines;
-                if (mags.length >= MAX_STRIPES || mags.length > currentLength || freed) {
-                    return true;
+                int mask = mags.length - 1;
+                int index = (int) (threadId & mask);
+                for (int i = 0, m = Integer.numberOfTrailingZeros(~mask); i < m; i++) {
+                    Magazine mag = mags[index + i & mask];
+                    if (buf == null) {
+                        buf = mag.newBuffer();
+                    }
+                    if (mag.tryAllocate(size, maxCapacity, buf, reallocate)) {
+                        // Was able to allocate.
+                        return buf;
+                    }
                 }
-                int preferredChunkSize = mags[0].sharedPrefChunkSize;
-                Magazine[] expanded = new Magazine[mags.length * 2];
-                for (int i = 0, l = expanded.length; i < l; i++) {
-                    Magazine m = new Magazine(this);
-                    m.localPrefChunkSize = preferredChunkSize;
-                    m.sharedPrefChunkSize = preferredChunkSize;
-                    expanded[i] = m;
+                expansions++;
+            } while (expansions <= EXPANSION_ATTEMPTS && tryExpandMagazines(mags.length));
+
+            // The magazines failed us; contention too high and we don't want to spend more effort expanding the array.
+            return null;
+        }
+
+        long usedMemory() {
+            long sum = 0;
+            for (Chunk chunk : centralQueue) {
+                sum += chunk.capacity();
+            }
+            if (threadLocalMagazine != null) {
+                sum += threadLocalMagazine.usedMemory.get();
+            } else {
+                for (Magazine magazine : magazines) {
+                    sum += magazine.usedMemory.get();
                 }
-                magazines = expanded;
-            } finally {
-                magazineExpandLock.unlockWrite(writeLock);
             }
-            for (Magazine magazine : mags) {
-                magazine.free();
+            return sum;
+        }
+
+        private boolean tryExpandMagazines(int currentLength) {
+            if (currentLength >= MAX_STRIPES) {
+                return true;
+            }
+            final Magazine[] mags;
+            long writeLock = magazineExpandLock.tryWriteLock();
+            if (writeLock != 0) {
+                try {
+                    mags = magazines;
+                    if (mags.length >= MAX_STRIPES || mags.length > currentLength || freed) {
+                        return true;
```

Comment:
Shouldn't this be false ?
```suggestion
                        return false;
```

Reply:
`tryExpandMagazines` only returns a boolean to allowed it to participate in the loop condition, and it always returns `true` to let the calling loop retry…